Added Hazelnut Stat quest, Longhouse options and removed 2h damroll bonus

airworthytick opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
- removed 2h damroll bonus (commented out)
- added framework for hazelnut (archdruid) stat quest
- removed base 10 spirit as it is now chosen with hazelnut choice
- added dropdown for hazelnut choice
- added new longhouse choices
- slightly modified display of longhouse choices to show stat bonuses.  This was done as new choices include a - option

## Motivation and Context
The changes now allow people to build lists that can accommodate the new stat quest choices without resorting to janky backdoor options like using the aux slots.  Also resolves the 2h damroll bonus which is no longer a thing and was removed from LegendMUD.

## How Has This Been Tested?
Tested in a test docker environment.  Tested all new drop down options and old ones to verify everything works properly.  Also tested saving in both new lists and an imported list from the current live hub.  Was unable to see any issues.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

steventhorne commented on www/src/public/js/controllers/builder/main.js:
```diff
@@ -1517,7 +1541,10 @@
                         fromStatQuests += 3;
                     }
-                    if ($scope.selectedList.baseStats.amulet == 1) {
+                    if ($scope.selectedList.baseStats.amulet == 1 || $scope.selectedList.baseStats.hazelnut == 1) {
```
if hazelnut == 1 and amulet != 1, it'll still increase this stat by 20, because of the or in the first if

airworthytick replied:
Good catch, I had caught this error and corrected it originally but either this one snuck past me or I forgot to save.

Fixed!

steventhorne commented on www/src/public/js/controllers/builder/main.js:
```diff
@@ -1912,7 +1963,7 @@
 
             // apply total bonuses
-            total += getStatTotalBonus(statName, total);
+            // total += getStatTotalBonus(statName, total);
```
I'm guessing this was a mistake?

steventhorne replied:
oh I see why you did this, I would prefer if you removed the "dam" case in the function, rather than commenting out the call to the function completely

airworthytick replied:
Corrected.  I decided to comment out the "dam" case instead of deletion like we talked about.  I figured it hurts nothing and leaves a bit of a cookie crumb for someone coming along later to add something if needed as a use case.

steventhorne commented on www/src/public/js/controllers/builder/main.js:
```diff
@@ -232,6 +238,9 @@
             newList.baseStats.amulet = Number(each[0]);
             each.shift();
 
+            newList.baseStats.hazelnut = Number(each[0]);
```
There are some issues with the way you're saving and loading the hazelnut stuff, we should talk

airworthytick replied:
I believe these are all resolved based on our conversations.  There is now new version checking and testing was done for both importing and exporting of v2 and v3 lists, as well as loading v2 and v3 cookies.

steventhorne reviewed (changes requested):
Overall this is really great for a first PR!
There are a few issues, a couple you can fix on your own and one that I'll step you through.

Nice job!

steventhorne commented on www/src/public/js/controllers/builder/main.js:
```diff
@@ -304,6 +313,12 @@
                 baseStats.amulet = encoder.toNumber(listStr.slice(0,1));
             listStr = listStr.substring(1);
 
+            if (statList[0] === '_')
```
wrap this in a check for v3 lists

airworthytick replied:
Version check included as if statement for hazelnut value.

steventhorne reviewed (commented):
Make sure you update `localStorage.setItem("cl2", listCookieStr);` to cl3

steventhorne reviewed (commented):
Make sure to update the beginning of the `loadCharacterLists` function to get v3 first

airworthytick commented:
I believe I have resolved all saving issues.  I attempted to go with your suggestion of using the indexList value of being > 0 to throw errors to force it to version check but this failed due to I believe it being a while loop...so it always ended at 0.  I instead ended up using scope.slotOrder.length matching as if if things are coming from v2 they fail this check due to the hazelnut value pushing them up by 1.  In theory this could help with future versions as well but who knows.  Let me know if there are failure points here I haven't considered or if there is a better way to go about it but from what I can tell it works so I called it good for now.

I also added as a last minute change more optional slots.  People are adding buff stats to things like KSM values and breaking lists, giving more option lists should in theory allow them to do whatever crazy shit they are doing without causing any new issues (hopefully).  I added code to add the slots to existing lists and new lists alike.  I did some mild testing, it appeared to all work but again let me know if you see issues.

Thanks again!  Bust out that red pen!  (also I think I updated the pull request properly, it looks like my new changes have uploaded but I'm not 100%, the site isn't super intuitive to me so let me know if I failed)

steventhorne commented on www/src/public/js/controllers/builder/main.js:
```diff
@@ -146,27 +147,34 @@
-                    if (!listCookieStr) {
-                        listCookieStr = $cookies.get("cl1");
-                    }
+                var listCookieStr = localStorage.getItem("cl3")
```
There's a lot of duplicate work happening here that's quite inefficient.

On this line you're getting cl3, and then not doing anything with it, you should immediately check `if (listCookieStr)` and set it to version 3 if it's successful. In each of the other if statements you're also doing double the work. For example, on line 153, you're getting cl2 to see if it exists, and then you're setting the listCookieStr on the next line, by getting it from storage again. `localStorage.getItem()` is not a free call, so when you get it the first time, make sure you save it to the listCookieStr to reuse it later.

Additionally, you have all of these in if statements, but not else statements. Meaning that even if there is a cl3 item, it will also load up the cl1 and cl2 items (twice). All in all, there are always 7 calls to `getItem` here, when there should be at minimum 1 (they have a cl3) and at maximum 4 (they still have a cl1 for some reason). You're also missing a check for `cl`, the list without a version. While highly unlikely, it's still possible that someone has this.

airworthytick replied:
I believe I have this resolved.  Tested and working with v2 and v3's and hopefully fulfilling your desires.  I'm not sure on the original version stuff...I may have some bad checks there so let me know.

steventhorne commented on www/src/public/js/controllers/builder/main.js:
```diff
@@ -146,27 +147,34 @@
                         switch (listVersion) {
-                            case 1:
-                                var newList = createListFromString(listStrs[i]);
+                            case 3:
```
since case 3 and 2 are the same you can combine them into one case:

```
switch (listVersion) {
  case 1:
    var newList = createListFromString(listStrs[i]);
    break;
  case 2:
  case 3:
    var newList = createListFromStringV2(listStrs[i], listVersion);
    break;
  default:
    break;
}
```

airworthytick replied:
If factorio has taught me anything its that optimization and efficiency are key.  Fixed.

steventhorne commented on www/src/public/js/controllers/builder/main.js:
```diff
@@ -313,11 +319,14 @@
+                    baseStats.hazelnut = -1;
+                else
+                    baseStats.hazelnut = encoder.toNumber(listStr.slice(0,1));
+                    listStr = listStr.substring(1);
```
be careful of the indenting here, while it doesn't pose an issue, it might be confusing to another reader. They might think that the line was intended to be inside of the else statement, which might lead them to putting brackets around lines 327-328 and that would break things.

airworthytick replied:
Duh moment on my part but good eye.  Fixed.

steventhorne commented on www/src/public/js/controllers/builder/main.js:
```diff
@@ -313,11 +319,14 @@
+                else
+                    baseStats.hazelnut = encoder.toNumber(listStr.slice(0,1));
+                    listStr = listStr.substring(1);
+            }
```
We should have an else statement here that sets the hazelnut to spirit if it's not a v3 list, just to maintain the current stats on old lists.

airworthytick replied:
I agree.  Completed.

steventhorne commented on www/src/public/js/controllers/builder/main.js:
```diff
@@ -351,6 +360,19 @@
+            }
+
+            // Version Check, if # of item slots loaded does not equal total number in current version it will throw an error
+            if(itemIndex > $scope.slotOrder.length) {
```
great idea!

steventhorne commented on www/src/public/js/controllers/builder/main.js:
```diff
@@ -1311,9 +1338,10 @@
                 }
                 else {
-                    selectListByIndex($scope.selectedListIndex);
+                    selectListByIndex($scope.allLists.length - 1);
```
What's the intended purpose for this change?

The old logic would make it so that if you delete a character and they're not at the end of your list of characters, it'll select the next character by default. This updated logic will make it so that it always selects the last character in your character list when you delete one

airworthytick replied:
I...have no idea.  I don't remember ever willfully making changes to the delete function.  I'll check the original and just move it back, no clue how that happened.

steventhorne reviewed (changes requested):
Getting closer!

steventhorne commented on www/src/public/js/controllers/builder/main.js:
```diff
@@ -144,38 +144,42 @@
+                            listVersion = 1;
+                        }
+                        else {
+                            listVersion = "cl";
```
this else statement should do a `localStorage.getItem("cl")`. As far as the `listVersion` goes, both `cl1` and `cl` are considered version 1, so you can set that outside of this specific if/else statement, before you try `getItem("cl1")`. This part is really weird and dated so I don't blame you for the confusion

airworthytick replied:
Fixed!

steventhorne replied:
This is still an issue because you're setting the listVersion to "cl", when it should be set to 1. You can just move the `listVersion = 1` out of the if/else since both the if and else are going to set it to 1, here.

steventhorne reviewed (changes requested):
so close! one tiny change that supports legacy stuff

steventhorne reviewed (changes requested):
stat updates look good, still a very minor issue with loading, though

steventhorne reviewed (changes requested):
The .env_example file was deleted? This should not be deleted

airworthytick commented:
Should be all resolved.  I accidently messed up the example file and deleted it locally, apparently it then committed it for some reason.  I have recreated it now.

steventhorne reviewed (approved):
Well done! All around, great changes. Next time, I'd like to see some more descriptive commit messages.
